cores/portduino/AsyncUDP.cpp

```diff
@@ -0,0 +1,131 @@
+#include "AsyncUDP.h"
+
+void _asyncudp_async_cb(uv_async_t *handle) {
+    AsyncUDP *udp = (AsyncUDP *)handle->data;
+    udp->_DO_NOT_CALL_async_cb();
+};
+
+AsyncUDP::AsyncUDP() {
+    _handler = NULL;
+    _connected = false;
+    uv_loop_init(&_loop);
+    _async.data = this;
+    uv_async_init(&_loop, &_async, _asyncudp_async_cb);
+};
+
+void _asyncudp_alloc_buffer_cb(uv_handle_t *handle, size_t suggested_size, uv_buf_t *buf) {
+    buf->base = (char *)malloc(suggested_size);
+    buf->len = suggested_size;
+};
+
+void _asyncudp_on_read_cb(uv_udp_t *handle, ssize_t nread, const uv_buf_t *buf, const struct sockaddr *addr, unsigned flags) {
+    // on the handle.data void pointer cast to AsyncUDP and call the _on_read function
+    AsyncUDP *udp = (AsyncUDP *)handle->data;
+    udp->_DO_NOT_CALL_uv_on_read(handle, nread, buf, addr, flags);
+}
+
+void AsyncUDP::_DO_NOT_CALL_uv_on_read(uv_udp_t *handle, ssize_t nread, const uv_buf_t *buf, const struct sockaddr *addr, unsigned flags) {
+    if (nread == 0) {
+        // no data
+        return;
+    }
+
+    _handlerMutex.lock();
+    auto h = _handler;
+    _handlerMutex.unlock();
+    if (h) {
+        AsyncUDPPacket packet((uint8_t*)buf->base, nread);
+        h(packet);
+    }
+    free(buf->base);
+};
+
+bool AsyncUDP::listenMulticast(const IPAddress addr, uint16_t port, uint8_t ttl) {
+    if (_connected) {
+        return false;
+    }
+    // FIXME: implement error handling rather than raising SIGSEGV
+    if (uv_udp_init(&_loop, &_socket) < 0) {
+        raise(SIGSEGV);
```

Comment:
This probably
https://github.com/meshtastic/framework-portduino/blob/e54db5e057f8f89b920e1f1e35a22380ec3ad153/cores/portduino/Utility.cpp#L12